src/wayland/wl.c

```diff
@@ -122,34 +140,52 @@ static void output_handle_scale(void *data, struct wl_output *wl_output,
         wake_up();
 }
 
+#ifdef WL_OUTPUT_NAME_SINCE_VERSION
+static void output_handle_name(void *data, struct wl_output *wl_output,
+                const char *name) {
+        struct dunst_output *output = data;
+        output->name = g_strdup(name);
+        LOG_D("Output global %" PRIu32 " name %s", output->global_name, name);
+}
+#endif
+
 static const struct wl_output_listener output_listener = {
         .geometry = output_handle_geometry,
         .mode = output_handle_mode,
         .done = noop,
         .scale = output_handle_scale,
+#ifdef WL_OUTPUT_NAME_SINCE_VERSION
+        .name = output_handle_name,
+        .description = noop,
+#endif
 };
 
-static void create_output( struct wl_output *wl_output, uint32_t global_name) {
+static void create_output(struct wl_registry *registry, uint32_t global_name, uint32_t version) {
         struct dunst_output *output = g_malloc0(sizeof(struct dunst_output));
         if (output == NULL) {
                 LOG_E("allocation failed");
                 return;
         }
 
+        uint32_t max_version = 3;
+#ifdef WL_OUTPUT_NAME_SINCE_VERSION
+        max_version = WL_OUTPUT_NAME_SINCE_VERSION;
+#endif
         bool recreate_surface = false;
         static int number = 0;
         LOG_I("New output found - id %i", number);
         output->global_name = global_name;
-        output->wl_output = wl_output;
+        output->wl_output = wl_registry_bind(registry, global_name, &wl_output_interface,
+                        CLAMP(version, 3, max_version));
```

Comment:
I think this can be done in a less confusing way. You're trying to select the version WL_OUTPUT_NAME_SINCE_VERSION if that's available and otherwise version 3, right? Why need the clamp?

Reply:
The version announced by the compositor can be lower than `WL_OUTPUT_NAME_SINCE_VERSION`. In this case it would not be correct to bind to it (IIRC, it's even a protocol error).

So we have a minimum supported version (3), maximum supported (`WL_OUTPUT_NAME_SINCE_VERSION` if compiled with the newer headers), and we should be able to accept anything within this range.

Reply:
I believe you are able to bind to a lower version than the compositor announces. Otherwise compositors would not be able to talk with older apps.
So my suggestion is to either bind to version 3 or the other required version. Not anything in between. I think that would make the code a little bit clearer

Reply:
So... `version = version >= max_version ? max_version : 3`? I don't believe that to be a significant improvement over `CLAMP(version, 3, max_version)`.

Besides, that works in the current case (where WL_OUTPUT_NAME_SINCE_VERSION is 4), but with other protocols we may benefit from additional events/requests introduced in the versions between min and max. Binding strictly to min or max would prevent that.
That's why, as a general pattern, clamping is better.

Reply:
>  That's why, as a general pattern, clamping is better.
Okay, I'll accept that argument